rbd.c

```diff
@@ -49,9 +64,254 @@ struct rbd_aio_cb {
+	rados_shutdown(state->cluster);
+
+	state->cluster = NULL;
+	state->io_ctx = NULL;
+	state->image = NULL;
+
+	pthread_spin_lock(&state->lock);
+	state->state = TCMU_RBD_CLOSED;
+	pthread_spin_unlock(&state->lock);
+}
+
+static int tcmu_rbd_image_open(struct tcmu_device *dev)
+{
+	struct tcmu_rbd_state *state = tcmu_get_dev_private(dev);
+	int ret;
+
+	pthread_spin_lock(&state->lock);
+	if (state->state == TCMU_RBD_OPENED) {
+		tcmu_dev_dbg(dev, "skipping open. Already opened\n");
+		pthread_spin_unlock(&state->lock);
+		return 0;
+	}
+
+	if (state->state != TCMU_RBD_CLOSED) {
+		tcmu_dev_dbg(dev, "skipping open. state %d\n", state->state);
+		pthread_spin_unlock(&state->lock);
+		return -EBUSY;
+	}
+	state->state = TCMU_RBD_OPENING;
+	pthread_spin_unlock(&state->lock);
+
+	/* TODO locking */
+	ret = rados_create(&state->cluster, NULL);
+	if (ret < 0) {
+		tcmu_dev_dbg(dev, "Could not create cluster. (Err %d)\n", ret);
+		goto set_closed;
+	}
+
+	/* Fow now, we will only read /etc/ceph/ceph.conf */
+	rados_conf_read_file(state->cluster, NULL);
+	rados_conf_set(state->cluster, "rbd_cache", "false");
+
+	ret = rados_connect(state->cluster);
+	if (ret < 0) {
+		tcmu_dev_err(dev, "Could not connect to cluster. (Err %d)\n",
+			     ret);
+		goto rados_shutdown;
+	}
+
+	ret = rados_ioctx_create(state->cluster, state->pool_name,
+				 &state->io_ctx);
+	if (ret < 0) {
+		tcmu_dev_err(dev, "Could not create ioctx for pool %s. (Err %d)\n",
+			     state->pool_name, ret);
+		goto rados_destroy;
+	}
+
+	ret = rbd_open(state->io_ctx, state->image_name, &state->image, NULL);
+	if (ret < 0) {
+		tcmu_dev_err(dev, "Could not open image %s. (Err %d)\n",
+			     state->image_name, ret);
+		goto rados_destroy;
+	}
+
+	pthread_spin_lock(&state->lock);
+	state->state = TCMU_RBD_OPENED;
+	pthread_spin_unlock(&state->lock);
+	return 0;
+
+rados_destroy:
+	rados_ioctx_destroy(state->io_ctx);
+	state->io_ctx = NULL;
+rados_shutdown:
+	rados_shutdown(state->cluster);
+	state->cluster = NULL;
+set_closed:
+	pthread_spin_lock(&state->lock);
+	state->state = TCMU_RBD_CLOSED;
+	pthread_spin_unlock(&state->lock);
+	return ret;
+}
+
+static int tcmu_rbd_image_reopen(struct tcmu_device *dev)
+{
+	int ret;
+
+	tcmu_dev_dbg(dev, "image reopen\n");
+	tcmu_rbd_image_close(dev);
+	ret = tcmu_rbd_image_open(dev);
+	tcmu_dev_dbg(dev, "image reopen got %d\n", ret);
+	return ret;
+}
+
+static int tcmu_rbd_lock(struct tcmu_device *dev)
+{
+	struct tcmu_rbd_state *state = tcmu_get_dev_private(dev);
+	int ret = 0, attempts = 0;
+
+	/*
+	 * TODO: Add retry/timeout settings to handle windows/ESX.
+	 * Or, set to transitioning and grab the lock in the background.
+	 */
+	while (attempts++ < 5) {
+		ret = is_exclusive_lock_owner(dev);
+		if (ret == 1) {
+			ret = 0;
+			break;
+		} else if (ret == -ESHUTDOWN) {
+			ret = tcmu_rbd_image_reopen(dev);
+			continue;
+		} else if (ret < 0) {
+			sleep(1);
+			continue;
+		}
+
+		ret = rbd_lock_acquire(state->image, RBD_LOCK_MODE_EXCLUSIVE);
```

Comment:
My original thinking is that you should execute ```rbd_lock_get_owners``` first -- if you receive -ENOENT, that's no problem since it just means that no one currently owns the lock. Since it's an exclusive lock, you can assume only a single owner will be returned. Next, assuming this method is invoked on an ALUA path state change, you would invoke ```rbd_lock_break``` providing the lock owner retrieved from ```rbd_lock_get_owners```. Finally, acquire the lock as you are currently doing. Since you broke the lock, if you have a failure acquiring the lock -- it would be a real issue. If possible, to prevent the possibility of a second ALUA failover, it would be safest if you can somehow validate the returned lock owner with the ALUA path before breaking the lock (unless you know you cannot have another failover between looking up the owner and breaking its lock).